visualizer.py

```diff
@@ -0,0 +1,323 @@
+
+        if self.trigger is not TriggerState.TRIGGERED:
+            return
+
+        if frame.format_id != self._current_format:
+            self.traces = dict()
+            self.times = np.empty(self.max_size, dtype=np.uint32)
+            self._current_format = frame.format_id
+
+        # Append trace data
+        self._append_traces(frame)
+
+        # Check for the capture buffer being filled.
+        if self.trace_index >= self.max_size:
+            logging.info('Got %d samples - stopping trigger', self.max_size)
+            self.trigger = TriggerState.STOPPED
+
+
+    def get_traces(self):
+        """ Get the traces and times to display in the graph. """
+        return (self.times - self.times[0]), self.traces
+
+
+class StreamVisualizer:
+    """ A visualizer for stabilizer's livestream data. """
+
+    def __init__(self, receiver: StreamReceiver, document: bokeh.document.Document):
+        """ Initialize the visualizer.
+
+        Args:
+            receiver: The StreamReceiver object for managing stream reception.
+            document: The Bokeh document to draw into.
+        """
+
+        self.receiver = receiver
+
+        figure = bokeh.plotting.figure(output_backend="webgl", sizing_mode='stretch_both')
+
+        # Add a trigger button
+        trigger_button = bokeh.models.Button(label='Single', button_type='primary')
+        trigger_button.on_click(self.capture)
+
+        force_button = bokeh.models.Button(label='Force', button_type='primary')
+        force_button.on_click(self.receiver.force_trigger)
+
+        self._capture_duration_input = bokeh.models.TextInput(title='Capture Duration',
+                                                              value='0.001',
+                                                              width=100,
+                                                              sizing_mode='fixed')
+        self._capture_duration_input.on_change('value', self.handle_duration)
+        self._trigger_state = bokeh.models.Div(text='Trigger State: <b>IDLE</b>')
+
+        control_layout = bokeh.layouts.column(self._trigger_state, trigger_button, force_button,
+                                              self._capture_duration_input)
+
+        self.layout = bokeh.layouts.row(figure, control_layout, sizing_mode='stretch_height')
+
+        document.theme = 'dark_minimal'
+        document.add_root(self.layout)
+
+        self.doc = document
+        self._callback = None
+
+        self._data_store = bokeh.models.ColumnDataSource(data={
+            'time': []
+        })
+
+
+    def handle_duration(self, _attr, old_value, new_value):
+        """ Handle updates to the capture duration input. """
+        try:
+            self._capture_duration_input.value = str(float(new_value))
+        except ValueError:
+            self._capture_duration_input.value = str(float(old_value))
+
+
+    def update(self):
+        """ Periodic check for capture status. """
```

Comment:
I'd like to remove as much polling action as possible and put push/asyncio in place instead.
Also the `add_periodic_callback` below.